File: src/main/java/cn/labzen/web/ap/evaluate/annotation/AbandonedEvaluator.java
package cn.labzen.web.ap.evaluate.annotation;

import cn.labzen.web.annotation.Abandoned;
import cn.labzen.web.ap.config.Config;
import cn.labzen.web.ap.internal.Utils;
import cn.labzen.web.ap.internal.element.ElementMethod;
import cn.labzen.web.ap.suggestion.DiscardSuggestion;
import cn.labzen.web.ap.suggestion.RemoveSuggestion;
import cn.labzen.web.ap.suggestion.Suggestion;
import com.squareup.javapoet.TypeName;

import java.util.List;
import java.util.Map;

public final class AbandonedEvaluator implements MethodErasableAnnotationEvaluator {

  private static final TypeName SUPPORTED = TypeName.get(Abandoned.class);

  @Override
  public boolean support(TypeName type) {
    return SUPPORTED.equals(type);
  }

  @Override
  public List<? extends Suggestion> evaluate(Config config, TypeName type, Map<String, Object> members) {
    RemoveSuggestion removeSuggestion = new RemoveSuggestion(Utils.getSimpleName(type), ElementMethod.class);
    DiscardSuggestion discardSuggestion = new DiscardSuggestion();
    return List.of(removeSuggestion, discardSuggestion);
  }
}

File: src/main/kotlin/cn/labzen/web/ap/evaluate/annotation/AbandonedEvaluator.kt
package cn.labzen.web.ap.evaluate.annotation

import cn.labzen.web.annotation.Abandoned
import cn.labzen.web.ap.config.Config
import cn.labzen.web.ap.internal.Utils
import cn.labzen.web.ap.internal.element.ElementMethod
import cn.labzen.web.ap.suggestion.Suggestion
import cn.labzen.web.ap.suggestion.impl.DiscardSuggestion
import cn.labzen.web.ap.suggestion.impl.RemoveSuggestion
import com.squareup.javapoet.TypeName

class AbandonedEvaluator : MethodErasableAnnotationEvaluator {

  override fun support(type: TypeName): Boolean =
    SUPPORTED == type

  override fun evaluate(config: Config, type: TypeName, members: Map<String, Any?>): List<Suggestion> =
    listOf(RemoveSuggestion(Utils.getSimpleName(type), ElementMethod::class.java), DiscardSuggestion())

  companion object {
    private val SUPPORTED = TypeName.get(Abandoned::class.java)
  }
}